docs/dev/writing-code.md

```diff
@@ -190,20 +199,270 @@ Don't use this pattern unless you're adding to a class where it's already in use
+*   The garbage collector cleans up the `EventLoopGroup`,
+    and the `PyCapsule` referenced by `EventLoopGroup._binding`.
+*   The `PyCapsule`'s destructor function runs, which looks something like:
+    ```C
+    static on_capsule_destroyed_fn(PyObject *capsule) {
+        struct event_loop_group_binding *binding = PyCapsule_GetPointer(capsule);
+
+        // release reference to underlying implementation
+        aws_event_loop_group_release(binding->native)
+
+        // free binding struct's memory
+        aws_mem_release(binding);
+    }
+    ```
+*   IF nothing else has a strong reference to `struct aws_event_loop_group`:
+    *   then it begins its shutdown process, and its memory is cleaned up when shutdown completes.
+*   ELSE something else has a strong reference to `struct aws_event_loop_group`:
+    *   so it won't begin its shutdown until the last reference is released.
+
+Note: In the past, the aws-c libraries didn't have reference counting for any C structs.
+You will still find older code in our Python bindings that tries to keep the entire dependency trees
+of Python objects alive via `Py_INCREF(x)` (TODO: remove needless complexity).
+You can't always look at existing code to see "the right way" of doing things.
+
+## A More Complex Example
+
+The sample above is simplified, it only shows Python calling into C.
+But `aws_event_loop_group` has a callback that fires when it finishes shutting down.
+That means C needs to call into Python AFTER the Python `EventLoopGroup` object has been cleaned up.
+
+For C to call into Python, it must reference a Python object
+(the function itself, or an object with a member function).
+This means our binding needs to store a strong reference and
+keep that Python object alive until the callback has fired.
+
+### The Wrong Way to Build it
+
+You MUST NOT create a reference cycle!
+
+You might be tempted to give `event_loop_group_binding` a strong reference to the `EventLoopGroup` instance.
+Then C could simply call a private member function like `EventLoopGroup._on_shutdown_complete()`.
+But this design creates a reference cycle (see image below):
+
+![Diagram of bad binding with reference cycle](./binding-bad-cycle.svg)
+
+### Option 1 - Pass Callbacks to C
+
+Most of our bindings work like this:
+
+![Diagram where C References callbacks](binding-callbacks.svg)
+
+Within `EventLoopGroup.__init__()` a "callable" is defined and passed down to C.
+`event_loop_group_binding` keeps a strong reference to this Python object.
+The code looks something like:
+```py
+    class EventLoopGroup:
+        def __init__(self, ...):
+
+            # define callable local function
+            def shutdown_callback():
+                ...do stuff...
+
+            self._binding = _awscrt.event_loop_group_new(shutdown_callback, ...)
+```
+
+When the final shutdown callback happens in C, the Python callable
+is invoked, and then the reference is released via `Py_DECREF(x)`.
```

Comment:
Might want to indicate that the callable that's passed to native gets Py_INCREF(x) to clarify keeping a strong reference there since we're pointing out the Py_DECREF(x) here.